product_attribute_variant_rules/models/product_attribute_rule.py

```diff
@@ -0,0 +1,152 @@
+        string="Available preconditions",
+    )
+
+    available_postcondition_attribute_ids = fields.Many2many(
+        comodel_name="product.attribute",
+        compute="_compute_available_postcondition_attribute_ids",
+        string="Available postconditions",
+    )
+
+    @api.depends(
+        "product_tmpl_id.attribute_line_ids",
+        "product_attribute_value_postcondition_ids",
+    )
+    def _compute_available_precondition_attribute_ids(self):
+        """
+        Compute the available preconditions.
+        """
+        for rule in self:
+            rule.available_precondition_attribute_ids = (
+                rule.product_tmpl_id.attribute_line_ids.mapped("attribute_id")
+                - rule.product_attribute_value_postcondition_ids.mapped("attribute_id")
+            )
+
+    @api.depends(
+        "product_tmpl_id.attribute_line_ids", "product_attribute_value_precondition_ids"
+    )
+    def _compute_available_postcondition_attribute_ids(self):
+        """
+        Compute the available postconditions.
+        """
+        for rule in self:
+            rule.available_postcondition_attribute_ids = (
+                rule.product_tmpl_id.attribute_line_ids.mapped("attribute_id")
+                - rule.product_attribute_value_precondition_ids.mapped("attribute_id")
+            )
+
+    def _is_combination_possible(self, combination):
+        """
+        Check if the combination is possible with the rules defined on the
+        product template.
+        """
+        # Check if the combination matches the preconditions
+        if not self._combination_matches_conditions(
+            combination, self.product_attribute_value_precondition_ids
+        ):
+            # If the combination does not match the preconditions,
+            # the rule is not applied
+            return True
+
+        # Check if the combination matches the postconditions
+        match = self._combination_matches_conditions(
+            combination, self.product_attribute_value_postcondition_ids
+        )
+
+        if self.type == "only" and match:
+            # Both conditions are met in only, the combination is possible
+            return True
+        elif self.type == "never" and not match:
+            # Precondition is met but postcondition is not met in never,
+            # the combination is possible
+            return True
+
+        # The combination is not possible
+        return False
+
+    def _combination_matches_conditions(self, combination, conditions):
+        """
+        Check if the combination matches the given conditions.
+        """
+        # If there is no condition, the combination matches
+        # (only possible for the preconditions)
+        if not conditions:
+            return True
+
+        # Check if the combination matches the preconditions ANDed between
+        # different attributes
+        for attribute, attribute_values in self._aggregate_conditions(
```

Comment:
You can use the groupby function of odoo

```
from odoo.tools import groupby


for attribute, values in groupby(combinations):


```

https://github.com/odoo/odoo/blob/53a3509a10f4ed2f4edf5374bb19ebda57e7530e/odoo/tools/misc.py#L1225